### ISSUE TITLE
[v2.5] numeric_binning with the frequency table is slow

### ISSUE THREAD

**Issue by kiedanski** [saulpw/visidata#1050]
First, thanks for Visidata, I really like the concept, and I'm glad someone took the time to develop it.

I have a dataset with 600k rows. Some of them are categorical and some are numeric.

For categorical columns (100~ unique items), the frequency table has an acceptable performance (is not fast, but not terribly slow). 
When trying to use the frequency table for numerical attributes, however, the performance degrades to the point that I need to close the application. Running `value_counts` using pandas from an IPython terminal is almost instantaneous.

I tried playing around the binning of numerical data but did not work :/

**Comment by anjakefala** [comment]
Hi @danski3456, thanks for the report! Can you please share your version of VisiData? It would also be helpful if you could share a minimal commandlog, through which this could be reproduced.

**Comment by anjakefala** [comment]
Additionally, if you could, it would be very helpful if you could attach 1000 random rows from that dataset! There could be something unique to the dataset's structure that is resulting in the dramatic performance drop.

**Comment by kiedanski** [comment]
The version running is: saul.pw/VisiData v2.5

It actually happens to me with any numerical column. It seems that it does not correctly bin the data, and it builds a LOT of buckets.

**Comment by kiedanski** [comment]
<img width="1430" alt="image" src="https://user-images.githubusercontent.com/49415614/126664334-154469e9-d4a5-4912-8580-4b25fcfba54e.png">

<img width="1102" alt="image" src="https://user-images.githubusercontent.com/49415614/126664369-6fa7db68-f161-4555-bb3a-196938800f26.png">

**Comment by kiedanski** [comment]
Tried the following script:

```python
import pandas as pd
import numpy as np

r = np.random.RandomState(1234)

A = r.uniform(-1000, 1000, size=(1000000, 10))

df = pd.DataFrame(A)
df.to_csv("test.csv")
```

And vd seems to perform faster in that dataset. Even then, it fails to bin data into ranges.

**Comment by anjakefala** [comment]
Hi danski!

The relevant information and bug is that you are using numeric binning and setting the histogram_bins! Numeric binning is being improved over time, but it is very possible that there are still edge cases.

Could you please supply us with a sample dataset that is resulting in the poor performance and awkward binning?

For now, I can try to debug using the pandas script you provided.

**Comment by kiedanski** [comment]
Unfortunately, I can't share the relevant data.

Is there some other way to get bins to work correctly?

**Comment by anjakefala** [comment]
If you turn numeric binning off, then the numbers get treated as categorical values.  

If you are sure that you cannot anonymise or randomise the context for the data, so that enough subset of the numbers can be shared to reproduce the problem, I can do my best to try to reproduce things, but failing that there might not be much that I can do. =( That is the pickle. Without any sort of data that reproduces the problem, I might not be able to help.

When I am addressing the issue, I will try with the pandas script, and some datasets I have.

**Comment by kiedanski** [comment]
I'll see what I can do. 

Regarding normal binning of numerical data. Is there a way at the moment to bin, for example, data into 10 ranges?

**Comment by anjakefala** [comment]
The technical way is to open the `Options Sheet` (`Shift+O`), set `options.numeric_binning` to True, and set `options.histogram_bins` to 10.

Then return to your sheet, and ensure the value you want to bin has a numeric type (`#`, `%`, etc) and press `Shift+F` on it.

I just tested this, and it works for my data, but I am not sure how things will look for you!

**Comment by kiedanski** [comment]
It works, but it takes quite a long time to calculate the bins

**Comment by jorgelaranjo** [comment]
Info on threads and Visidata [Threads](https://github.com/saulpw/visidata/blob/77b2c3ee73aeb16a25a9cc08eeff0c53800b6467/docs/api/async.rst
)![image](https://user-images.githubusercontent.com/1196202/131180231-aadb132c-97cf-4cbf-9b86-d8249275c821.png)

![image](https://user-images.githubusercontent.com/1196202/131175000-4556cdc0-da10-4aed-a19c-0fa55d4b8626.png)

**Comment by ghost** [comment]
Something to keep in mind here. pandas vs visidata is not exactly a fair test, because pandas uses a columnar data model, and builds on top of very fast numeric libraries written in C, while visidata uses a row based data model, and to my knowledge is written in pure Python.

**Comment by saulpw** [comment]
Merged into #1068.
